### PR TITLE
feat: add opt-in AIA certificate chain fetching (tlsAIAFetch)

### LINKED ISSUE #5941: k6 fails TLS verification for certificates requiring AIA chain fetching
### Summary

k6 fails TLS verification with x509: certificate signed by unknown authority for servers that rely on AIA (Authority Information Access) to complete their certificate chain, even when the certificate is valid and trusted by browsers.

### Background
Some servers (including those behind Cloudflare) present a leaf certificate without sending the full intermediate chain in the TLS handshake. Instead, they advertise a URL in the certificate's AIA extension field where the missing intermediate can be fetched. Browsers implement this natively,  they fetch the intermediate, complete the chain, and verify successfully.

Go's crypto/tls does not implement AIA fetching. As a result, k6 cannot build the trusted chain and rejects the certificate.

### Expected behaviour
k6 should fetch missing intermediate certificates via AIA URLs advertised in the certificate, build the full chain, and verify successfully.

### PR DESCRIPTION
**Pull request by vtorosyan**
## What?

Go's crypto/tls does not fetch missing intermediate certificates via the Authority Information Access (AIA) extension; browsers do. Servers that send only a leaf certificate and rely on AIA to fill in the chain cause k6 TLS handshakes to fail with "x509: certificate signed by unknown authority" even when the chain is technically valid.

This commit adds a new `tlsAIAFetch` option (env: K6_TLS_AIA_FETCH) that enables AIA chain fetching on a per-test basis. When enabled, k6 intercepts failed chain verification, follows IssuingCertificateURL fields (BFS, up to 5 hops), retries verification with any fetched intermediates, and then restores hostname verification — exactly what browsers do.

## Implementation details:
- lib/netext/aia.go: core WrapTLSConfigForAIAFetching function; uses a separate http.Client (not measured as test traffic) with a 10s timeout; process-level sync.Map cache so repeated handshakes pay the HTTP cost at most once per AIA URL; circular-reference protection via seen map
- lib/options.go: new TLSAIAFetch null.Bool option (default off)
- internal/js/runner.go: opt-in wiring — wraps the per-VU tls.Config when tlsAIAFetch: true; no-op when insecureSkipTLSVerify is also true
- internal/js/modules/k6/grpc/client.go: fix buildTLSConfig to copy VerifyPeerCertificate and VerifyConnection callbacks from the parent config, so AIA verification is not silently bypassed for gRPC connections that supply explicit TLS options (cacerts/cert/key)

Tests:
- lib/netext/aia_test.go: 10 parallel unit tests covering happy path, complete chain passthrough, disabled/plain-Go regression, InsecureSkip- Verify passthrough, unreachable URL, fetch timeout, HTTP error, malformed cert, circular AIA references, and hostname-mismatch enforcement
- internal/js/runner_aia_test.go: full VU integration test (Source and Archive runners) verifying the enabled/disabled toggle end-to-end

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests for my changes.
- [x] I have run linter and tests locally (`make check`) and all pass.

## Checklist: Documentation (only for k6 maintainers and if relevant)

**Please do not merge this PR until the following items are filled out.**

- [ ] I have added the correct milestone and labels to the PR.
- [ ] I have updated the release notes: _link_
- [ ] I have updated or added an issue to the [k6-documentation](https://github.com/grafana/k6-docs): grafana/k6-docs#NUMBER if applicable
- [ ] I have updated or added an issue to the [TypeScript definitions](https://github.com/grafana/k6-DefinitelyTyped/tree/master/types/k6): grafana/k6-DefinitelyTyped#NUMBER if applicable

## Related PR(s)/Issue(s)

Fixes https://github.com/grafana/k6/issues/5941

### CODE REVIEW

File: internal/js/modules/k6/browser/common/kill_other.go [hunk 1]
```diff
@@ -1,5 +1,4 @@
 //go:build !linux
-// +build !linux
```
Comment (vtorosyan) [inline internal/js/modules/k6/browser/common/kill_other.go:2]: The // +build !linux line was the old pre-1.17 syntax. When both are present in a file, they must agree and the //go:build one takes precedence — the old form is purely redundant. The modernize linter flagged it as dead weight for exactly this reason.

**Comment by vtorosyan** [comment]
Using an internal project to validate the changes manually as well: https://github.com/grafana/sm-k6-tls-test.

### Without a patch

```
════════════════════════════════════════════════
  Linux k6 TLS Test (reproduces AIA bug)
════════════════════════════════════════════════

  k6 image  : grafana/k6
  AIA net   : container:37c337d53d43781f25bf9d38661e9c2c04221d429abed080ca99c8ca0e4c7351

  On Linux/CGO_ENABLED=0, 'aia-chain' is expected to FAIL.
  All other scenarios should PASS.

  Scenarios
  ─────────────────────────────────────────
  aia-chain               FAIL ✗
  complete-chain          PASS ✓
  self-signed             PASS ✓
  self-signed/skip        PASS ✓
  insecure-skip           PASS ✓
  wrong-hostname          PASS ✓
  wrong-hostname/skip     PASS ✓

  ─────────────────────────────────────────

  Suite healthy: 6/7 passed.

  aia-chain FAILED — this is the AIA bug reproducing correctly on Linux.
  Apply the AIA-fetching patch to k6 and re-run to validate the fix.
  Use 'make compare-docker' to diff a patched build against this one.
```

### With patch applied (from this branch)

```
════════════════════════════════════════════════════════
  Linux k6 TLS Comparison (CGO_ENABLED=0, no macOS caveat)
════════════════════════════════════════════════════════

  Image A:     grafana/k6 (v2.0.0-20260424105549-2a014daaa9f4+dirty (commit/2a014daaa9-dirty, go1.26.1, linux/arm64))
  Image B:     k6-patched (v2.0.0-rc1.0.20260506132255-99b68176e5a3 (commit/99b68176e5, go1.25.9, linux/arm64))
  AIA net:     container:12b7fdca9e2f

  On Linux/CGO_ENABLED=0, aia-chain FAILS on unpatched k6 (Image A)
  and should PASS on the patched build (Image B).

  Scenario                  Image A                 Image B                 Assessment
  ──────────────────────────────────────────────────────────────────────────
  aia-chain                 FAIL (expected) ✓  PASS ✓             Behaviour changed as expected
  complete-chain            PASS ✓             PASS ✓             Consistent (both correct)
  self-signed               PASS ✓             PASS ✓             Consistent (both correct)
  self-signed/skip          PASS ✓             PASS ✓             Consistent (both correct)
  insecure-skip             PASS ✓             PASS ✓             Consistent (both correct)
  wrong-hostname            PASS ✓             PASS ✓             Consistent (both correct)
  wrong-hostname/skip       PASS ✓             PASS ✓             Consistent (both correct)

  ──────────────────────────────────────────────────────────────────────────

  All scenarios match expected behaviour.

  AIA chain fetch:
  Image A (unpatched): aia-chain FAIL — Go TLS ignores AIA (bug reproduced on Linux)
  Image B (patched): aia-chain PASS — patch working correctly.
```

**Comment by mstoykov** [comment]
Hi are you aware of a not built by us system that we can test this against - not for unit test but just for sanity checking. So that we are not just testing our implementation against our implementation, but it isn't what actually happens in the "wild"

**Comment by vtorosyan** [comment]
@mstoykov thanks for the idea!
I found the [`incomplete-chain.badssl.com`](https://incomplete-chain.badssl.com), which seems to be a good fit. 

It's a TLS test service maintained by the Chrome security team that intentionally omits the intermediate certificate from the handshake, requiring AIA fetching to build a valid chain. It uses a real publicly-trusted certificate, so there's no custom CA involved.

I added a scenario to [the test repo](https://github.com/grafana/sm-k6-tls-test) that hits it directly with `tlsAIAFetch: true`.

```bash
# unpatched grafana/k6 on Linux (CGO_ENABLED=0) — fails on a real public cert
docker run --rm \
  -v "$PWD/scenarios:/scenarios:ro" \
  -e K6_TLS_AIA_FETCH=true \
  grafana/k6 run --quiet /scenarios/badssl-aia/test.js
# → x509: certificate signed by unknown authority

# same command with the patched binary
docker run --rm \
  -v "$PWD/scenarios:/scenarios:ro" \
  -e K6_TLS_AIA_FETCH=true \
  k6-patched run --quiet /scenarios/badssl-aia/test.js
# → PASS — TLS verified against incomplete-chain.badssl.com
```

`make compare-docker` now also prints a second table for the third-party result alongside the local scenarios:

```
  Third-party validation (incomplete-chain.badssl.com)

  Scenario       Image A (unpatched)     Image B (patched)     Assessment
  ─────────────────────────────────────────────────────────────────────────
  badssl-aia     FAIL (expected) ✓       PASS ✓                Behaviour changed as expected
```

Full instructions and the scenario script are in the test repo: https://github.com/grafana/sm-k6-tls-test

**Comment by vtorosyan** [comment]
What I _have not_ tested yet is the performance/latency impact of the change (which worries me most tbh). Chatting with @mem, he mentioned that his observed latency impact was very minimum, given that the AIA fetching is more of an edge-case scenario, and we are caching locally -- I don't expect to see any observable difference.

That said, working currently to add a benchmark test as well. Is there perhaps a metric etc we can emit to monitor as well?

**Comment by vtorosyan** [comment]
## Benchmark tests

I have added few benchmark scenarios and run the tests on local. 

Three scenarios were benchmarked:

| Scenario | Description |
|---|---|
| **Baseline** | `tlsAIAFetch` disabled. Server sends a complete certificate chain. Normal Go TLS behaviour. |
| **AIA Warm Cache** | `tlsAIAFetch` enabled. Server sends an incomplete chain, but k6 already fetched the missing certificate on a previous connection and has it in memory. No extra HTTP request needed. |
| **AIA Cold Cache** | `tlsAIAFetch` enabled. Server sends an incomplete chain and k6 has never seen this server before — it must make an extra HTTP request to fetch the missing certificate. |

Each benchmark was run 3 times. All tests ran on loopback (local machine talking to itself).

### Baseline

| Run | Time per connection | Memory per connection | Allocations |
|---|---|---|---|
| 1 | 709,794 ns (≈ 0.71 ms) | 129,802 B (≈ 127 KB) | 1,006 |
| 2 | 684,390 ns (≈ 0.68 ms) | 129,742 B (≈ 127 KB) | 1,006 |
| 3 | 705,398 ns (≈ 0.71 ms) | 129,776 B (≈ 127 KB) | 1,006 |

### AIA Warm Cache

| Run | Time per connection | Memory per connection | Allocations |
|---|---|---|---|
| 1 | 706,479 ns (≈ 0.71 ms) | 130,585 B (≈ 128 KB) | 1,055 |
| 2 | 719,501 ns (≈ 0.72 ms) | 130,574 B (≈ 128 KB) | 1,055 |
| 3 | 715,380 ns (≈ 0.72 ms) | 130,573 B (≈ 128 KB) | 1,055 |

### AIA Cold Cache

| Run | Time per connection | Memory per connection | Allocations |
|---|---|---|---|
| 1 | 615,518 ns (≈ 0.62 ms) | 141,120 B (≈ 138 KB) | 1,174 |
| 2 | 629,506 ns (≈ 0.63 ms) | 141,138 B (≈ 138 KB) | 1,174 |
| 3 | 616,393 ns (≈ 0.62 ms) | 141,136 B (≈ 138 KB) | 1,174 |

## Summary

| Scenario | Avg time/connection | Allocations | vs Baseline |
|---|---|---|---|
| Baseline | ~700 µs | 1,006 | — |
| AIA Warm Cache | ~714 µs | 1,055 | +14 µs, +49 allocs |
| AIA Cold Cache | ~620 µs | 1,174 | noise, +168 allocs |

## What the numbers mean

### Timing

The time differences between all three scenarios are within ~10% of each other and fall within normal measurement noise. This is expected: these tests run on loopback (no real network), so the dominant cost is **crypto computation**, not networking. On a real network with real latency, the cold cache scenario would **show a visible overhead**, one extra HTTP round-trip to fetch the missing certificate, typically **10–100 ms** depending on where the CA server is.

### Allocations (the reliable signal)

The allocation count is stable across runs

- **Warm cache adds 49 allocations** over baseline. This is the overhead of k6 running its own certificate verification logic instead of Go's built-in. This is negligible.
- **Cold cache adds 168 allocations** over baseline. This comes from making an HTTP request to fetch the missing certificate, parsing it, and storing it in the cache.

### Caching behaviour

The cold cache cost is a **one-time cost per unique server** (keyed by AIA URL). After the first connection, every subsequent VU connecting to the same server hits the warm cache path, paying only the 49-allocation overhead, not a full HTTP round-trip. In a typical k6 test with hundreds or thousands of VUs all hitting the same target, only the very first connection triggers a fetch.

## Conclusion

The `tlsAIAFetch` feature has **negligible overhead in steady state** (warm cache). The cold cache cost - a single HTTP request to a CA server, is a one-time expense per target host for the lifetime of the test run, and is only relevant when the feature is explicitly opted into via `tlsAIAFetch: true` and follows the AIA path. **So opting in to tlsAIAFetch on a well-configured server carries essentially zero cost.**

**Comment by mstoykov** [comment]
@vtorosyan can you please rebase this

**Comment by vtorosyan** [comment]
@mstoykov I apologize, completely missed the comment about rebasing and the branch got stale. I will do this ASAP.

**Comment by vtorosyan** [comment]
@mstoykov and team, sorry for getting back late with the merge. I have missed the ask/comment. Since my changes were created before we required signed commits, I had to create a new PR unfortunately (as all my rebase attempts failed, I could not make the branch up to date in a clean way). 

Here is the PR https://github.com/grafana/k6/pull/6137

I appreciate the help and understanding!
